File: sdk/python/src/conduyt/core/wire.py
"""CONDUYT Wire Format — Packet encode/decode"""

from __future__ import annotations

import struct
from dataclasses import dataclass

from .constants import PROTOCOL_VERSION, MAGIC, HEADER_SIZE
from .crc8 import crc8


@dataclass
class ConduytPacket:
    version: int
    type: int
    seq: int
    payload: bytes


def wire_encode(packet: ConduytPacket) -> bytes:
    """Encode a ConduytPacket into raw wire bytes."""
    payload_len = len(packet.payload)
    buf = bytearray(HEADER_SIZE + payload_len)

    # MAGIC
    buf[0] = MAGIC[0]
    buf[1] = MAGIC[1]
    # VER
    buf[2] = packet.version
    # TYPE
    buf[3] = packet.type
    # SEQ
    buf[4] = packet.seq
    # LEN (little-endian uint16)
    struct.pack_into("<H", buf, 5, payload_len)
    # PAYLOAD
    buf[7 : 7 + payload_len] = packet.payload
    # CRC8 over [VER..end of PAYLOAD]
    crc_region = buf[2 : 7 + payload_len]
    buf[7 + payload_len] = crc8(crc_region)

    return bytes(buf)


def wire_decode(data: bytes | bytearray) -> ConduytPacket:
    """Decode raw wire bytes into a ConduytPacket. Raises ValueError on error."""
    if len(data) < HEADER_SIZE:
        raise ValueError(f"Incomplete packet: need {HEADER_SIZE} bytes, got {len(data)}")

    if data[0] != MAGIC[0] or data[1] != MAGIC[1]:
        raise ValueError(f"Invalid magic: 0x{data[0]:02x}{data[1]:02x}")

    version = data[2]
    if version != PROTOCOL_VERSION:
        raise ValueError(f"Version mismatch: expected {PROTOCOL_VERSION}, got {version}")

    pkt_type = data[3]
    seq = data[4]
    payload_len = struct.unpack_from("<H", data, 5)[0]

    total = HEADER_SIZE + payload_len
    if len(data) < total:
        raise ValueError(f"Incomplete packet: need {total} bytes, got {len(data)}")

    crc_region = data[2 : 7 + payload_len]
    expected_crc = crc8(crc_region)
    actual_crc = data[7 + payload_len]

    if expected_crc != actual_crc:
        raise ValueError(f"CRC mismatch: expected 0x{expected_crc:02x}, got 0x{actual_crc:02x}")

    payload = bytes(data[7 : 7 + payload_len])
    return ConduytPacket(version=version, type=pkt_type, seq=seq, payload=payload)


def make_packet(pkt_type: int, seq: int, payload: bytes = b"") -> ConduytPacket:
    """Create a ConduytPacket ready for encoding."""
    return ConduytPacket(version=PROTOCOL_VERSION, type=pkt_type, seq=seq, payload=payload)

File: sdk/swift/Sources/ConduytKit/Core/Wire.swift
import Foundation

/// Protocol constants
public enum ConduytProtocol {
    public static let version: UInt8 = 0x01
    public static let magic: [UInt8] = [0x43, 0x44]
    public static let headerSize = 8
}

/// Decoded CONDUYT packet
public struct ConduytPacket {
    public let version: UInt8
    public let type: UInt8
    public let seq: UInt8
    public let payload: Data

    public init(version: UInt8 = ConduytProtocol.version, type: UInt8, seq: UInt8, payload: Data = Data()) {
        self.version = version
        self.type = type
        self.seq = seq
        self.payload = payload
    }
}

/// Wire format errors
public enum WireError: Error, LocalizedError {
    case incompletePacket
    case invalidMagic
    case versionMismatch
    case crcMismatch(expected: UInt8, actual: UInt8)

    public var errorDescription: String? {
        switch self {
        case .incompletePacket: return "Incomplete packet"
        case .invalidMagic: return "Invalid magic bytes"
        case .versionMismatch: return "Protocol version mismatch"
        case .crcMismatch(let e, let a): return "CRC mismatch: expected 0x\(String(e, radix: 16)), got 0x\(String(a, radix: 16))"
        }
    }
}

/// Encode a ConduytPacket into raw wire bytes.
public func wireEncode(_ pkt: ConduytPacket) -> Data {
    let payloadLen = pkt.payload.count
    var buf = Data(capacity: ConduytProtocol.headerSize + payloadLen)

    buf.append(contentsOf: ConduytProtocol.magic)
    buf.append(pkt.version)
    buf.append(pkt.type)
    buf.append(pkt.seq)
    buf.append(UInt8(payloadLen & 0xFF))
    buf.append(UInt8((payloadLen >> 8) & 0xFF))
    buf.append(pkt.payload)

    let crcRegion = Array(buf[2..<(7 + payloadLen)])
    buf.append(CRC8.compute(crcRegion))

    return buf
}

/// Decode raw wire bytes into a ConduytPacket.
public func wireDecode(_ data: Data) throws -> ConduytPacket {
    guard data.count >= ConduytProtocol.headerSize else {
        throw WireError.incompletePacket
    }

    guard data[0] == ConduytProtocol.magic[0] && data[1] == ConduytProtocol.magic[1] else {
        throw WireError.invalidMagic
    }

    let version = data[2]
    guard version == ConduytProtocol.version else {
        throw WireError.versionMismatch
    }

    let type = data[3]
    let seq = data[4]
    let payloadLen = Int(data[5]) | (Int(data[6]) << 8)

    let total = ConduytProtocol.headerSize + payloadLen
    guard data.count >= total else {
        throw WireError.incompletePacket
    }

    let crcRegion = Array(data[2..<(7 + payloadLen)])
    let expected = CRC8.compute(crcRegion)
    let actual = data[7 + payloadLen]

    guard expected == actual else {
        throw WireError.crcMismatch(expected: expected, actual: actual)
    }

    let payload = data[7..<(7 + payloadLen)]
    return ConduytPacket(version: version, type: type, seq: seq, payload: Data(payload))
}
